allennlp/models/coreference_resolution/coref.py

```diff
@@ -0,0 +1,824 @@
+            the start and end word indices of the top spans that survived the pruning stage.
+        antecedent_indices : ``torch.IntTensor``
+            A tensor of shape ``(num_spans_to_keep, max_antecedents)`` representing for each top span
+            the index (with respect to top_spans) of the possible antecedents the model considered.
+        predicted_antecedents : ``torch.IntTensor``
+            A tensor of shape ``(batch_size, num_spans_to_keep)`` representing, for each top span, the
+            index (with respect to antecedent_indices) of the most likely antecedent. -1 means there
+            was no predicted link.
+        loss : ``torch.FloatTensor``, optional
+            A scalar loss to be optimised.
+        """
+        # Shape: (batch_size, document_length, embedding_size)
+        text_embeddings = self._lexical_dropout(self._text_field_embedder(text))
+
+        # Shape: (batch_size, document_length)
+        text_mask = util.get_text_field_mask(text).float()
+
+        # Shape: (batch_size, num_spans, 1)
+        span_mask = (span_starts >= 0).float()
+        # IndexFields return -1 when they are used as padding. As we do
+        # some comparisons based on span widths when we attend over the
+        # span representations that we generate from these indices, we
+        # need them to be <= 0. This is only relevant in edge cases where
+        # the number of spans we consider after the pruning stage is >= the
+        # total number of spans, because in this case, it is possible we might
+        # consider a masked span.
+        span_starts = F.relu(span_starts.float()).long()
+        span_ends = F.relu(span_ends.float()).long()
+
+        # Shape: (batch_size, num_spans, embedding_size)
+        span_embeddings = self._compute_span_representations(text_embeddings,
+                                                             text_mask,
+                                                             span_starts,
+                                                             span_ends)
+        # Compute mention scores.
+        # Shape: (batch_size, num_spans, 1)
+        mention_scores = self._mention_scorer(self._mention_feedforward(span_embeddings))
+        mention_scores += span_mask.log()
+
+        # Prune based on mention scores.
+        num_spans_to_keep = int(math.floor(self._spans_per_word * text_embeddings.size(1)))
+
+        # Shape: (batch_size, num_spans_to_keep)
+        # These are indices (with values between 0 and num_spans) into
+        # the span_embeddings tensor.
+        top_span_indices = self._prune_and_sort_spans(mention_scores, num_spans_to_keep)
+
+        # Shape: (batch_size * num_spans_to_keep)
+        # torch.index_select only accepts 1D indices, but here
+        # we need to do select spans for each element in the batch.
+        # This reformats the indices to take into account their index into the batch.
```

Comment:
Adding something like "We precompute this here to make our multiple calls to `util.batched_index_select` below more efficient." might be helpful, so the reader doesn't wonder why you get what seems like duplicate information passed to `batched_index_select` in both `top_span_indices` and `flat_top_span_indices`.

Reply:
Also, it'd be nice to have some kind of note like this up above, after the `top_span_indices = ...` line: "Now that we've decided which spans are actually mentions, we reformat all of our variables to be in terms of `num_spans_to_keep` instead of `num_spans`, so we don't waste computation on spans that we've already discarded."